packages/compiler-cli/src/ngtsc/hmr/src/extract_locals.ts

```diff
@@ -0,0 +1,228 @@
+    // to runtime code. Intentionally excludes interfaces and type declarations.
+    if (
+      ts.isClassDeclaration(node) ||
+      ts.isFunctionDeclaration(node) ||
+      (ts.isEnumDeclaration(node) &&
+        !node.modifiers?.some((m) => m.kind === ts.SyntaxKind.ConstKeyword))
+    ) {
+      if (node.name) {
+        results.add(node.name.text);
+      }
+      continue;
+    }
+
+    // Variable declarations.
+    if (ts.isVariableStatement(node)) {
+      for (const decl of node.declarationList.declarations) {
+        trackBindingName(decl.name, results);
+      }
+      continue;
+    }
+
+    // Import declarations.
+    if (ts.isImportDeclaration(node) && node.importClause) {
+      const importClause = node.importClause;
+
+      // Skip over type-only imports since they won't be emitted to JS.
+      if (importClause.isTypeOnly) {
+        continue;
+      }
+
+      // import foo from 'foo'
+      if (importClause.name) {
+        results.add(importClause.name.text);
+      }
+
+      if (importClause.namedBindings) {
+        const namedBindings = importClause.namedBindings;
+
+        if (ts.isNamespaceImport(namedBindings)) {
+          // import * as foo from 'foo';
+          results.add(namedBindings.name.text);
+        } else {
+          // import {foo} from 'foo';
+          namedBindings.elements.forEach((el) => {
+            if (!el.isTypeOnly) {
+              results.add(el.name.text);
+            }
+          });
+        }
+      }
+      continue;
+    }
+  }
+
+  return results;
+}
+
+/**
+ * Adds all the variables declared through a `ts.BindingName` to a set of results.
+ * @param node Node from which to start searching for variables.
+ * @param results Set to which to add the matches.
+ */
+function trackBindingName(node: ts.BindingName, results: Set<string>): void {
+  if (ts.isIdentifier(node)) {
+    results.add(node.text);
+  } else {
+    for (const el of node.elements) {
+      if (!ts.isOmittedExpression(el)) {
+        trackBindingName(el.name, results);
+      }
+    }
+  }
+}
+
+/**
+ * Visitor that will traverse an AST looking for potential top-level variable reads.
+ * The reads are "potential", because the visitor doesn't account for local variables
+ * inside functions.
+ */
+class PotentialTopLevelReadsVisitor extends o.RecursiveAstVisitor {
+  readonly allReads = new Set<string>();
+
+  override visitReadVarExpr(ast: o.ReadVarExpr, context: any) {
+    this.allReads.add(ast.name);
+    super.visitReadVarExpr(ast, context);
+  }
+
+  override visitWrappedNodeExpr(ast: o.WrappedNodeExpr<unknown>, context: any) {
+    if (this.isTypeScriptNode(ast.node)) {
+      this.addAllTopLevelIdentifiers(ast.node);
+    }
+
+    super.visitWrappedNodeExpr(ast, context);
+  }
+
+  /**
+   * Traverses a TypeScript AST and tracks all the top-level reads.
+   * @param node Node from which to start the traversal.
+   */
+  private addAllTopLevelIdentifiers = (node: ts.Node) => {
```

Comment:
nit: can we declare this function as a class method?
```suggestion
  private addAllTopLevelIdentifiers(node: ts.Node) {
```

Reply:
I had to define it as an arrow function, because it uses `this` and is being passed in directly in the `forEachChild` call below.